compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -1857,12 +1857,40 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                 }
             }
             GeneratorInteriorOrUpvar::Upvar(upvar_span) => {
+                // `Some(ref_ty)` if `target_ty` is `&T` and `T` fails to impl `Sync`
+                let refers_to_non_sync = match target_ty.kind() {
+                    ty::Ref(_, ref_ty, _) => match self.evaluate_obligation(&obligation) {
+                        Ok(eval) => {
+                            if eval.may_apply() {
+                                None
+                            } else {
+                                Some(ref_ty)
+                            }
+                        }
+                        _ => None,
+                    },
+                    _ => None,
+                };
                 let mut span = MultiSpan::from_span(upvar_span);
-                span.push_span_label(
-                    upvar_span,
-                    format!("has type `{}` which {}", target_ty, trait_explanation),
-                );
-                err.span_note(span, &format!("captured value {}", trait_explanation));
+                match refers_to_non_sync {
+                    Some(ref_ty) => {
+                        span.push_span_label(
+                            upvar_span,
+                            format!(
+                                "has type `{}` -- refers to `{}` which is not `Sync`",
+                                target_ty, ref_ty
+                            ),
```

Comment:
```suggestion
                            format!(
                                "has type `{}` which is not `Send`, because `{}` is not `Sync`",
                                target_ty,
                                ref_ty,
                            ),
```